input/postgres/server_stats.go

```diff
@@ -21,12 +21,64 @@ SELECT
 FROM pg_catalog.pg_control_checkpoint()
 `
 
+const xminHorizonSQL string = `
+WITH slots AS (
+	SELECT
+		xmin,
+		catalog_xmin
+	FROM pg_replication_slots
+	WHERE xmin IS NOT NULL
+	ORDER BY age(xmin) DESC
+	LIMIT 1
+)
+SELECT
+(
+	SELECT
+		backend_xmin
+	FROM pg_stat_activity
+	WHERE backend_xmin IS NOT NULL OR backend_xid IS NOT NULL
+	ORDER BY greatest(age(backend_xmin), age(backend_xid)) DESC
+	LIMIT 1
+) as backend,
+(
+	SELECT xmin FROM slots
+) as replication_slot_xmin,
+(
+	SELECT catalog_xmin FROM slots
+) as replication_slot_catalog_xmin,
+(
+	SELECT
+		transaction AS xmin
+	FROM pg_prepared_xacts
+	ORDER BY age(transaction) DESC
+	LIMIT 1
+) as prepare_xact,
+(
+	SELECT
+		backend_xmin
+	FROM pg_stat_replication
+	WHERE backend_xmin IS NOT NULL
+	ORDER BY age(backend_xmin) DESC
+	LIMIT 1
+) as standby
+`
+
 func GetServerStats(logger *util.Logger, db *sql.DB, postgresVersion state.PostgresVersion, systemType string) (state.PostgresServerStats, error) {
 	var stats state.PostgresServerStats
 	var transactionIdSQL string
 
-	// Only collect xact ID related stats with non-replicas
+	// Only collect transaction ID or xmin horizon related stats with non-replicas
 	if isReplica, err := getIsReplica(db); err == nil && !isReplica {
+		// Query xmin horizon before querying the current transaction ID
+		// as the backend_xmin from pg_stat_activity can point to the "next" transaction ID.
```

Comment:
[note for reviewer] This behavior is interesting, but apparently, when you do `select * from pg_stat_activity`, that backend will get the "next" current transaction ID as `backend_xmin`.
This means, if there is no other queries having backend_xmin or backend_xid, the result of `xmin_horizion_backend` becomes the next current transaction ID.
I wasn't able to find out why this happens in the code (though I wasn't looking hard), but I was able to continuously reproduce so I think this order is the safe thing to have.
(If it's the other order, the `xmin_horizion_backend` will be sent as a negative value as it's bigger than the current transaction ID)

Reply:
I think I started understanding why this happen. `backend_xmin` is always the "next" current transaction ID because all transaction IDs less than xmin are either committed and visible, or rolled back and dead. Meaning, xmin is the ID that is _not_ committed. When there is no other transaction is running, the next ID will be the one that is not committed.

This is a nature of xmin, so tweaking the order is kind of a workaround of this, but I think it's good enough so I'll leave it as is.